### DISCUSSION TITLE
Spotless — Diagnose and Check

### DISCUSSION THREAD

**Discussion by x80486** [diffplug/spotless#2332] [Q&A] [answered]
When exploring the tasks for a Gradle project, the relevant tasks for Spotless are:

```shell
$ ./gradlew tasks
...

Verification tasks
------------------
check - Runs all checks.
spotlessApply - Applies code formatting steps to sourcecode in-place.
spotlessCheck - Checks that sourcecode satisfies formatting steps.
spotlessDiagnose
spotlessJava
spotlessJavaApply
spotlessJavaCheck
spotlessJavaDiagnose
spotlessKotlinGradle
spotlessKotlinGradleApply
spotlessKotlinGradleCheck
spotlessKotlinGradleDiagnose

...
```

`spotlessApply` is clear, and so is `spotlessCheck`; however, `spotlessDiagnose` lacks a description. Is `spotlessDiagnose` really necessary? Does `spotlessCheck` also run all the checks and verifications that `spotlessDiagnose` performs?

I'm trying to build a pipeline to check for formatting errors based on the Spotless configuration in question, but I'm unsure if `spotlessDiagnose` makes sense for a CI/CD workflow.

It looks like `spotlessCheck` is intended to verify the formatting, while `spotlessDiagnose` may provide some information/diagnostics/insights into formatting issues?

**Comment by nedtwigg** [comment, marked as answer]
Some formatter steps have idempotence problems. Spotless fixes them automatically. Sometimes they cannot be fixed, in which case Spotless yells and says "run spotlessDiagnose to see why this is failing"

https://github.com/diffplug/spotless/blob/main/PADDEDCELL.md
